cmd/tailscale/main.go

```diff
@@ -858,8 +1059,108 @@ func (a *App) processUIEvents(w *app.Window, events []UIEvent, act jni.Object, s
 		case SearchEvent:
 			state.query = strings.ToLower(e.Query)
 			a.updateState(act, state)
+		case FileSendEvent:
+			a.sendFiles(e, files)
+		}
+	}
+}
+
+func (a *App) sendFiles(e FileSendEvent, files []File) {
+	go func() {
+		var totalSize int64
+		for _, f := range files {
+			totalSize += f.Size
 		}
+		if totalSize == 0 {
+			totalSize = 1
+		}
+		var totalSent int64
+		progress := func(n int64) {
+			totalSent += n
+			e.Updates(FileSendInfo{
+				State:    FileSendTransferring,
+				Progress: float64(totalSent) / float64(totalSize),
+			})
+			a.invalidate()
+		}
+		defer a.invalidate()
+		for _, f := range files {
+			if err := a.sendFile(e.Context, e.Target, f, progress); err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+				log.Printf("file send failed: %v", err)
+				e.Updates(FileSendInfo{
+					State: FileSendFailed,
+				})
+				return
+			}
+		}
+		e.Updates(FileSendInfo{
+			State: FileSendComplete,
+		})
+	}()
+}
+
+func (a *App) invalidate() {
+	select {
+	case a.invalidates <- struct{}{}:
+	default:
+	}
+}
+
+func (a *App) sendFile(ctx context.Context, target *apitype.FileTarget, f File, progress func(n int64)) error {
+	var body io.Reader
+	switch f.Type {
+	case FileTypeText:
+		body = strings.NewReader(f.Text)
+	case FileTypeURI:
+		f, err := a.openURI(f.URI, "r")
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		body = f
+	default:
+		panic("unknown file type")
```

Comment:
panic? does returning an error work instead?

Reply:
The default branch indicates a programming error; we're in control of the possible enum values.